### ISSUE TITLE
Regression in 3.6.0: running coverage combine after pytest-cov now errors out

### ISSUE THREAD

**Issue by njsmith** [pytest-dev/pytest-cov#222]
We run coverage in `parallel=True` mode, because we have some tests that spawn subprocesses, and we want to capture coverage for all of them. To get a complete report for codecov, the CI does:

```
pytest --cov ...
coverage combine
bash <(curl -s https://codecov.io/bash)
```

As far as I know, this was mandatory on pytest-cov 2.5.1 and earlier. Since 2.6.0 came out yesterday, our tests have been failing because `coverage combine` says `No data to combine` and it errors out, failing the test suite.

I *think* this may be happening because pytest-cov is now effectively running `coverage combine` itself, so that instead of being mandatory to run `coverage combine`, it's now mandatory to *not* to run `coverage combine`? Possibly because of changes in #178? For now that [seems to be working](https://github.com/python-trio/trio/pull/647) for us. But it'd be nice to confirm that our understanding is correct, and I figured you'd rather have a heads-up than not.

Our bug: https://github.com/python-trio/trio/issues/646

**Comment by ionelmc** [comment]
Yeah sorry, it was an unforeseen consequence of fixing the combining problems at the end of the run (#130). Unfortunately coverage 4.3 made `coverage combine` strict - https://github.com/nedbat/coveragepy/issues/412

Perhaps we could persuade @nedbat to revert that behavior, or have a config switch for it?

**Comment by ionelmc** [comment]
Maybe we could save the coverage file with a suffix as a workaround (then combine wouldn't fail). For how many ppl is this regression a problem?

**Comment by nedbat** [comment]
@ionelmc Hello!  I'm also struggling with how to make 2.6.0 work for me.  The issue says that now pytest-cov is doing it's own combine?  But it seems to not be picking up the `[paths]` setting? Issue #130 seems to be talking about a similar problem.

**Comment by ionelmc** [comment]
@nedbat ooof ... 2.6.0 was supposed to fix #130 - do you have an example where 2.6.0 don't work as expected?

**Comment by nedbat** [comment]
As you might expect, this is in a large distributed test suite.  I've worked around it by adding `mv .coverage .coverage.1; coverage combine` before my `coverage html` command.

I would like to get a small example of the behavior though, and I'm still working on that.

**Comment by blueyed** [comment]
Is the combining meant to happen always now?

I'm a bit confused that the combining is not being done for me with 2.6.0..

.coveragerc:

```ini
[run]
parallel = true
source = .
branch = true
[report]
include = pytest_django/*,pytest_django_test/*,tests/*
```

tox.ini:
```ini
setenv =
    PYTHONPATH = {toxinidir}:{env:PYTHONPATH:}
    COV_CORE_SOURCE={toxinidir}
    COV_CORE_CONFIG={toxinidir}/.coveragerc
    COV_CORE_DATAFILE={toxinidir}/.coverage.eager
passenv = PYTEST_ADDOPTS
```

I set `PYTEST_ADDOPTS='--cov-append --cov-report=xml --cov-report=term-missing:skip-covered'`

I have all of the `.coverage.eager.*` files afterwards, and have to run `coverage combine` like before.

(I am investigating why pytest-cov 2.6.0 causes coverage to drop on codecov for pytest-django by 74% (https://github.com/pytest-dev/pytest-django/pull/650))

**Comment by ionelmc** [comment]
Reverting the combining would make these tests fail:

- test_central_with_path_aliasing
- test_subprocess_with_path_aliasing

@nedbat so I've looked over the code in coverage 4.5.2 - it looks like the problem lies in the fact that neither of these methods applies aliasing to the current data:
- coverage.control.Coverage.stop
- coverage.control.Coverage.save
- coverage.control.Coverage.combine - only applies aliasing to data on disk (as .coverage files), not to the current data (self.data)

So if you want this change in pytest-cov reverted then please advise of a way to correctly apply aliasing to current data, or fix the Coverage object to apply full aliasing in either of those methods.

**Comment by cjw296** [comment]
This auto-combine is problematic when you have several runs of pytest (say across different versions of python) and want to combine them. Please at least add an option to prevent pytest-cov doing this combining.

**Comment by ionelmc** [comment]
@cjw296  `--cov-append` should solve that problem. If not then explain in more detail your usecase.

**Comment by cjw296** [comment]
@ionelmc - I don't think it will. I run CI in docker containers, one for each python version, with `parallel=True` in the coverage config. So, each container *should* spit out a bunch of `.coverage.whatever` files, these are written to a volume outside the container. A separate job then runs `coverage combine`, `coverage report`, etc on the whole set of files.

Because each run with `pytest-cov` ends up producing only a `.coverage` file, the workflow fails. In my case, it's easy enough to just not use `pytest-cov`, but wanted to let you know...

**Comment by pawamoy** [comment]
I'm having a similar issue. I run pytest with coverage multiple times and want to combine all results. If I don't use `--cov-append`, then between each run the `.coverage` file is deleted and then re-combined. At the end I get the coverage for the last run only. If I use `--cov-append`, now I have to explicitly delete the `.coverage` file myself before the runs.

In both cases it's one more command to run for me, but I honestly prefer combining explicitly myself, it feels less magic :slightly_smiling_face:  (I had to run experiments to be sure of what happens with or without `--cov-append`).

**Comment by ionelmc** [comment]
Would a config solution as in https://github.com/pytest-dev/pytest-cov/pull/435 (setting a COVERAGE_FILE env var) solve the combining problems here?

**Comment by cjw296** [comment]
Some that that could be put in `.coveragerc` or `pytest.ini` would be ideal.

**Comment by jiridanek** [comment]
My take on this is that while pytest users tend to naturally gravitate to pytest-cov, it is actually not necessary and running coverage.py directly works just as well, and the whole setup has less moving parts https://coverage.readthedocs.io/en/stable/#quick-start. So, unless you _need_ pytest-cov to achieve something specific, I suggest not to use it "just because we use pytest".
